src/main/java/org/veupathdb/service/eda/ss/model/reducer/StreamIntersectMerger.java

```diff
@@ -1,92 +1,131 @@
-    this.streams = sortedStreams.stream()
+    // If no input streams are provided, it should act as an "empty" Iterator.
+    // If any input stream is "empty", should act as an "empty" Iterator since we are intersecting Iterators.
+    if (sortedStreams.isEmpty() || !sortedStreams.stream().allMatch(Iterator::hasNext)) {
+      hasStarted = true;
+      peekableIdIndexStreams = null;
+      nextOutputIndex = null;
+      return;
+    }
+    // Convert iterators into peekable iterators.
+    this.peekableIdIndexStreams = sortedStreams.stream()
         .map(PeekableIterator::new)
         .toArray(PeekableIterator[]::new);
-    this.currentStream = Arrays.stream(streams)
-        .max(Comparator.comparing(iter -> iter.peek()))
-        .get();
-    this.currentStreamIdx = IntStream.range(0, streams.length)
-        .filter(i -> streams[i] == currentStream)
-        .findFirst()
-        .getAsInt();
-    // Cache a pointer to the nextStream to avoid excessive modulo operations.
-    this.nextStream = IntStream.range(0, streams.length)
-        .map(i -> (i + 1) % streams.length)
-        .toArray();
+    this.streamRing = new RingLinkedList(peekableIdIndexStreams);
+    this.currentIdIndexStream = streamRing.cursor.currentStream;
     hasStarted = false;
   }
 
   @Override
   public boolean hasNext() {
     if (!hasStarted) {
-      setCurrentElement();
+      findNextMatchingIdIndex();
       hasStarted = true;
     }
-    if (currentElement == null) {
-      return false;
-    } else {
-      return true;
-    }
+    return nextOutputIndex != null;
   }
 
   @Override
   public Long next() {
     if (!hasStarted) {
-      setCurrentElement();
+      findNextMatchingIdIndex();
       hasStarted = true;
     }
-    Long next = currentElement;
-    currentStream.next();
-    setCurrentElement();
+    Long next = nextOutputIndex;
+    currentIdIndexStream.next();
+    findNextMatchingIdIndex();
     return next;
   }
 
+  private void findNextMatchingIdIndex() {
+    boolean matchFound = false;
+    while (!matchFound && currentIdIndexStream.hasNext()) {
+      Long candidateIdIndex = currentIdIndexStream.peek();
+      // Iterate through each stream in the stream ring.
+      for (int i = 0; i < streamRing.size(); i++) {
+        currentIdIndexStream = streamRing.advanceCursor().currentStream;
+        // Advance the stream until it matches the candidate. If we overshoot, break and start our loop over
+        // with the ID index we found as our new candidate.
+        Long idIndex = currentIdIndexStream.skipUntilMatchesOrExceeds(candidateIdIndex);
+        if (!Objects.equals(idIndex, candidateIdIndex)) {
+          break;
+        }
+        // If we've iterated through all the streams without breaking, indicate a match was found.
+        if (i == streamRing.size() - 1) {
+          nextOutputIndex = candidateIdIndex;
+          matchFound = true;
+        }
+      }
+    }
+    if (!currentIdIndexStream.hasNext()) {
```

Comment:
I think you may be hiding a final match here.  matchFound gets set to true on line 76, breaking the loop.  But if that is the very last value in currentIdIndexStream, nextOutputIndex is set to null, deleting the match we just found.  Maybe just want to return on 77 in the "found" case.

Reply:
Since you went to the work to write it, I just went ahead with your implementation. I'll try to come up with a test-case for this as well.